Add SDS011

puuu opened the pull request:
## Description:


**Related issue (if applicable):** 

**Pull request in [esphome](https://github.com/esphome/esphome) with python changes (if applicable):** esphome/esphome#467
**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#185

## Checklist:

  - [x] The code change is tested and works locally.
  - [x] The code change follows the [standards](https://esphome.io/guides/contributing.html#contributing-to-esphome-core)

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

puuu committed:
Implement SDS011 sensor component

puuu committed:
Integrate SDS011

puuu committed:
use utf-8 symbol in units

OttoWinter commented on src/esphome/sensor/sds011_component.cpp:
```diff
@@ -0,0 +1,258 @@
+  ESP_LOGD(TAG,
+           "Got PM2.5 Concentration: %.1f µg/m³, PM10.0 Concentration: %.1f "
+           "µg/m³ from device 0x%04x",
+           pm_2_5_concentration, pm_10_0_concentration, device_id);
```
Can there be more than one device connected to one bus? If so, we should split add a device_id option so that users can set up multiple instances of this class and select which sensor to read from.

puuu replied:
It is not directly mentioned that multiple sensors are possible. However, since the data frames in the protocol contain a device id and a broadcast id (`0xffff`) exists, one would think that this is possible. But for me is not clear how possible collision will be handled on the bus. I have only one device and can not test it.

IMHO, to support multiple sensors on the bus, we need to split the component into two parts (similar to the dallas component):
1. Hub component to configure the uart port and handle incoming frames.
2. Sensor component with device_id (default to broadcast id) .

I had this in mind (and basically like that idea), but thought it was too over-engineered, especially since it is not clear if multiple sensors are possible.

OttoWinter replied:
Ok, I guess especially since we support SW UART on the ESP8266s there's not a real reason to support more than one device per bus. Even if we could, the sensors could run into trouble when they both hold the bus high/low at the same time.

OttoWinter replied:
Let's just remove that from the logs then, since the normal user shouldn't be concerned with this.

puuu replied:
Done

puuu committed:
remove unnecessary defines

puuu committed:
imporve logging

puuu commented on src/esphome/sensor/sds011_component.cpp:
```diff
@@ -0,0 +1,258 @@
+  const uint32_t now = millis();
+  if ((now - this->last_transmission_ >= 500) && this->data_index_) {
+    // last transmission too long ago. Reset RX index.
+    ESP_LOGD(TAG, "last transmission too long ago. Reset RX index.");
```
I changed this to verbose.

puuu commented on src/esphome/sensor/sds011_component.cpp:
```diff
@@ -0,0 +1,258 @@
+      this->data_index_ = 0;
+    } else if (!*check) {
+      // wrong data
+      ESP_LOGD(TAG, "Wrong data.");
```
I changed this to verbose and improved the message.

puuu commented on src/esphome/sensor/sds011_component.cpp:
```diff
@@ -0,0 +1,258 @@
+  uint8_t update_interval_minutes = this->get_update_interval() / 1000.0f / 60.0f;
+  this->set_update_interval((uint32_t)update_interval_minutes * 1000 * 60);
+  if (update_interval_minutes > 30) {
+    ESP_LOGW(TAG, "Update interval is longer than 30 min.");
```
According to the [protocol description](https://nettigo.pl/attachments/415), the interval should be between 1-30 minutes. I did not tested what will happen with higher values.

OttoWinter replied:
Ok then, that's ok.

But we should move this warning then to `dump_config` - setup is called very early on and unless the user is connected via USB they won't see the warning.

puuu replied:
Ok, I moved it to `dump_config()` and added an additional check in python.

puuu commented on src/esphome/sensor/sds011_component.cpp:
```diff
@@ -0,0 +1,258 @@
+  this->sds011_write_command_(command_data);
+
+  uint8_t update_interval_minutes = this->get_update_interval() / 1000.0f / 60.0f;
+  this->set_update_interval((uint32_t)update_interval_minutes * 1000 * 60);
```
The working period should be between 1-30 minutes and is given in minutes. (0 means continuous measurements). `update_interval_minutes` is sent to the sensor.  `update_interval` is only used to poll the sensor in query mode. . Casting in Python would be alright, but I do not know how to restrict the time type (in interval and maximum). Do you have an example?

OttoWinter replied:
Ok, it's rather simple (I'll add it to the esphome pr directly). We should remove the code here, components shouldn't set their own properties - it probably works now but it could cause some trouble later.

puuu replied:
Thank you! I removed the casting.

OttoWinter committed:
Merge branch 'dev' into sds011

OttoWinter committed:
Fix merge

Messed that one up

OttoWinter committed:
Run clang-format/clang-tidy

See https://github.com/esphome/esphome-core/pull/540

OttoWinter commented on src/esphome/sensor/sds011_component.cpp:
```diff
@@ -0,0 +1,243 @@
+  if (this->get_query_mode()) {
+    this->update();
+    this->set_timeout("delayed-update", 35000, [this]() {
+      this->set_interval("update", this->get_update_interval(), [this]() { this->update(); });
```
We should not override the PollingComponent's update `set_interval`.

What is the reason for this delayed second update?

puuu replied:
> We should not override the PollingComponent's update set_interval.

What do you mean with override? This component does not inherit from the `PollingComponent` class. However, with `query_mode`, it mimics the `PollingComponent` class. If you don't like it, I will change it.

> What is the reason for this delayed second update?

The query command only ask for the last measurement value and do not force a measurement. For working periods (update_interval >= 1min), the sensor need 30s for the measurement (mainly to get fresh air from the fan). So for a power cycle of esp and sensor, we have to wait 30s to get a valid value. In the case, the esp rebooted or was in deepsleep (esp8266), we can read a valid value with the first `update()`.

OttoWinter replied:
> What do you mean with override? This component does not inherit from the PollingComponent class. However, with query_mode, it mimics the PollingComponent class. If you don't like it, I will change it.

Oh, I assumed it was overriding `PollingComponent`. In that case we should override `PollingComponent` and `return` in update if we're not in query mode.

> The query command only ask for the last measurement value and do not force a measurement. For working periods (update_interval >= 1min), the sensor need 30s for the measurement (mainly to get fresh air from the fan). So for a power cycle of esp and sensor, we have to wait 30s to get a valid value. In the case, the esp rebooted or was in deepsleep (esp8266), we can read a valid value with the first update().

If the user only sees this on the first time they plug the device in I think it's not a problem. Deep sleep is likely also not that important here because the sensor surely uses tons of power that would quickly drain all batteries.

puuu replied:
As far as I can see, when we switch to `PollingComponent`, we can not delay the first query and always get the value of the previous measurement (which is several minutes old). IMHO, `query_mode` only makes sense for deep_sleep

OttoWinter replied:
That is true, however:

- `deep_sleep` does not really make sense with this integration I think. I'm pretty sure the sensor uses tons of energy even when in idle mode (might be wrong about this). And when it's in non-idle mode it surely uses a ton of energy too - so you'd only get at most a week of deep sleep uptime
- While delaying the initial reading might be the correct way of doing it, I think it's not that important, because: It only ever happens when the user power-cycles the setup or on initial data - IMO those two cases happen so seldomly that it's ok if we do not handle those correctly

puuu committed:
do not cast update interval

puuu committed:
do not log the device id

OttoWinter committed:
Merge branch 'dev' into sds011

OttoWinter committed:
Move SDS011 outside of ESP32-only define

OttoWinter reviewed (approved):
Looks good now, change `Component` to `PollingComponent` and it's ready to merge.

puuu committed:
better checking of float values

puuu committed:
remove query_mode

OttoWinter reviewed (approved):
LGTM! Thanks :+1:

OttoWinter committed:
Minor Changes

- Convert to PollingComponent (not strictly necessary, but if we decide to use update in the future this is better)
- Remove update_interval methods.
- Remove update_interval longer than 30min warning (takes up flash space and python already catches that)
- Lint line 50 (narrowing conversion from float to uint8_t)

OttoWinter committed:
Re-add RX-only mode

The logic here got removed in the query_mode removal, re-adding it

OttoWinter committed:
Use update interval in minutes

Not using PollingComponent was right. Also as the only way update interval is used here is minutes constrain the value to minutes.

OttoWinter commented:
Ok, so you were right about not using PollingComponent - it just doesn't fit very well here.

I realized that the update interval is only in minutes anyway with this sensor (since query mode was removed). So I think it is now cleaner to just have a `update_interval_min_` property that denotes the update interval in minutes. Then the python part also gets a bit cleaner.

The effect of this is also that to enable the continuous-measurement mode, the user needs to type `update_interval: 0min`, but I think that's ok (all values from 0s to 59s were essentially the same, so having one value is cleaner)

If you're ok with these changes, I'd be happy to merge. Also, please tell me if you prefer me not touching your git branches, I just think this way the whole conversation part and potential misunderstandings is removed - code talks :D

puuu commented:
OK. Tested and works with my last commit in the python part.

Thank you.
Fell free to merge!
